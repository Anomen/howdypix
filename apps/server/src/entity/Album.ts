import { Column, ViewEntity } from "typeorm";

@ViewEntity({
  expression: `
SELECT
   P.DIR,
   P.SOURCE,
   P.PARENTDIR,
   COUNT(DISTINCT P2.ID) AS NBPHOTOS,
   COUNT(DISTINCT P3.DIR) AS NBALBUMS,
   (
      SELECT
         FILE 
      FROM
         PHOTO P4 
      WHERE
         P4.DIR = P.DIR 
         AND P4.SOURCE = P.SOURCE LIMIT 0,
         1
   )
   AS PREVIEW 
FROM
   PHOTO P 
   LEFT JOIN
      PHOTO P2 
      ON (P2.DIR = P.DIR 
      AND P2.SOURCE = P.SOURCE) 
   LEFT JOIN
      PHOTO P3 
      ON (P3.PARENTDIR = P.DIR 
      AND P3.SOURCE = P.SOURCE) 
GROUP BY
   P.DIR,
   P.SOURCE,
   P.PARENTDIR;
    `,
})
export class Album {
  @Column("text")
  dir: string;

  @Column("text")
  source: string;

  @Column("text")
  parentDir: string;

  @Column("text")
  preview: string;

  @Column("int")
  nbPhotos: number;

  @Column("int")
  nbAlbums: number;
}
