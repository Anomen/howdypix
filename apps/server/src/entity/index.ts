export * from "./Source";
export * from "./Album";
export * from "./Photo";
export * from "./Search";
export * from "./SearchResult";
