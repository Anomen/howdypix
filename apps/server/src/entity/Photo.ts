import {
  Column,
  Entity,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  Unique,
} from "typeorm";
import { SearchResult } from "./SearchResult";
import { Album } from "./Album";

export enum PHOTO_STATUS {
  NOT_PROCESSED = "not_processed",
  PROCESSED = "processed",
}

@Entity()
@Unique(["album", "file"])
export class Photo {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  status: PHOTO_STATUS;

  @Column("int")
  inode: number;

  @Column("int")
  mtime: number;

  @Column("int")
  ctime: number;

  @Column("int")
  birthtime: number;

  @Column("int")
  size: number;

  @Column("text", { nullable: true })
  make: string;

  @Column("text", { nullable: true })
  model: string;

  @Column("int", { nullable: true })
  ISO: number;

  @Column("int", { nullable: true })
  shutter: number;

  @Column("int", { nullable: true })
  processedShutter: number;

  @Column("int", { nullable: true })
  aperture: number;

  @Column("int", { nullable: true })
  processedAperture: number;

  @Column("int", { nullable: true })
  createDate: number;

  // TODO to remove
  @Column("text")
  source: string;

  // TODO to remove
  @Column("text")
  parentDir: string;

  // TODO to remove
  @Column("text")
  dir: string;

  @ManyToOne(() => Album, (album) => album.photos)
  public album!: Album;

  @Column("text")
  file: string;

  @OneToMany(() => SearchResult, (searchResult) => searchResult.photo)
  public searchResults!: SearchResult[];
}
