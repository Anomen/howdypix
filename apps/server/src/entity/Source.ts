import {
  Column,
  Entity,
  getConnection,
  Not,
  OneToMany,
  PrimaryGeneratedColumn,
  Unique,
} from "typeorm";
import { Photo as EntityPhoto } from "./Photo";
import { Album } from "./Album";

@Entity()
@Unique(["source"])
export class Source {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  source: string;

  @Column("text")
  dir: string;

  @OneToMany(() => Album, (album) => album.sourceLk)
  public albums!: Album[];

  async getNbPhotos(): Promise<number> {
    const photoRepository = getConnection().getRepository(EntityPhoto);

    const { count } = await photoRepository
      .createQueryBuilder()
      .select("COUNT(distinct file)", "count")
      .where({
        source: this.source,
        dir: "",
        file: Not(""),
      })
      .getRawOne();

    return count;
  }

  async getNbAlbums(): Promise<number> {
    const photoRepository = getConnection().getRepository(EntityPhoto);

    const { count } = await photoRepository
      .createQueryBuilder()
      .select("COUNT(distinct dir)", "count")
      .where({
        source: this.source,
        dir: Not(""),
      })
      .getRawOne();

    return count;
  }

  async getPreview(): Promise<string> {
    const photoRepository = getConnection().getRepository(EntityPhoto);

    const data = await photoRepository
      .createQueryBuilder()
      .select("file", "preview")
      .where({
        source: this.source,
        file: Not(""),
      })
      .getRawOne();

    // TODO change for undefined picture
    return data?.preview || "TOTO.JPG";
  }

  static async upsert(sourceName: string, dir: string): Promise<void> {
    const sourceRepository = getConnection().getRepository(Source);
    const sourceToUpdate = await sourceRepository.findOne({
      source: sourceName,
    });

    if (sourceToUpdate) {
      // Update the value
      sourceToUpdate.dir = dir;
      await sourceRepository.save(sourceToUpdate);
    } else {
      const source = new Source();
      source.source = sourceName;
      source.dir = dir;
      await sourceRepository.save(source);
    }
  }
}
