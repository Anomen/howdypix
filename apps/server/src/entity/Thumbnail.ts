import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  PrimaryColumn,
  ManyToOne
} from "typeorm";
import { Photo } from "./Photo";

@Entity()
export class Thumbnail {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  path: string;

  @Column("text")
  dataSourceName: string;

  @Column("int")
  width: number;

  @Column("int")
  height: number;

  // @ManyToOne(type => Photo, photo => photo.thumbnails)
  // photo: Photo;
}
