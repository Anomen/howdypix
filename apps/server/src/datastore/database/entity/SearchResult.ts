import {
  Column,
  Entity,
  PrimaryGeneratedColumn,
  ManyToOne,
  CreateDateColumn,
  UpdateDateColumn,
} from "typeorm";
import { Search } from "./Search";
import { Photo } from "./Photo";

@Entity()
export class SearchResult {
  @PrimaryGeneratedColumn()
  id: number;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column("int")
  searchId!: number;

  @Column("int")
  photoId!: number;

  @Column("int")
  public order!: number;

  @ManyToOne(() => Search, (search) => search.searchResults)
  public search!: Search;

  @ManyToOne(() => Photo, (photo) => photo.searchResults)
  public photo!: Photo;
}
