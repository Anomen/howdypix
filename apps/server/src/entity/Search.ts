import { Column, Entity, PrimaryGeneratedColumn, OneToMany } from "typeorm";
import { SearchResult } from "./SearchResult";

@Entity()
export class Search {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: "text", nullable: true })
  orderBy: string;

  @Column({ type: "text", nullable: true })
  filterBy: string;

  @Column({ type: "text", nullable: true })
  album: string;

  @Column({ type: "text", nullable: true })
  source: string;

  @OneToMany(
    () => SearchResult,
    searchResult => searchResult.search
  )
  public searchResults!: SearchResult[];
}
