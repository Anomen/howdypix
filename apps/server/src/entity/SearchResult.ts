import { Column, Entity, PrimaryGeneratedColumn, ManyToOne } from "typeorm";
import { Search } from "./Search";
import { Photo } from "./Photo";

@Entity()
export class SearchResult {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("int")
  searchId!: number;

  @Column("int")
  photoId!: number;

  @Column("int")
  public order!: number;

  @ManyToOne(
    () => Search,
    search => search.searchResults
  )
  public search!: Search;

  @ManyToOne(
    () => Photo,
    photo => photo.searchResults
  )
  public photo!: Photo;
}
