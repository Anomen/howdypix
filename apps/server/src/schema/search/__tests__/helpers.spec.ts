import { Repository } from "typeorm";
import {
  doSearch,
  fetchSearchResult,
  findSavedSearch,
  saveNewSearch,
  saveSearchResult
} from "../searchHelpers";
import { Photo as EntityPhoto } from "../../../entity/Photo";
import { SearchResult as EntitySearchResult } from "../../../entity/SearchResult";
import { Search as EntitySearch } from "../../../entity/Search";

describe("searchHelpers", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test("findSavedSearch should make the right query", async () => {
    const findOne = jest.fn(() => ({ data: "yay" }));
    const repository = {
      findOne
    };

    expect(
      await findSavedSearch(
        (repository as unknown) as Repository<EntitySearch>,
        {
          orderBy: "DATE_ASC",
          source: "source",
          album: "album"
        }
      )
    ).toMatchSnapshot();

    expect(findOne.mock.calls).toMatchSnapshot();
  });

  test("saveNewSearch should make the right query", async () => {
    const save = jest.fn(() => ({ data: "yay" }));
    const repository = {
      save
    };

    expect(
      await saveNewSearch((repository as unknown) as Repository<EntitySearch>, {
        orderBy: "DATE_ASC",
        source: "source",
        album: "album"
      })
    ).toMatchSnapshot();

    expect(save.mock.calls).toMatchSnapshot();
  });

  test("doSearch should make the right query", async () => {
    const find = jest.fn(() => [{ data: "yay" }]);
    const repository = {
      find
    };

    expect(
      await doSearch((repository as unknown) as Repository<EntityPhoto>, {
        orderBy: "DATE_ASC",
        source: "source",
        album: "album"
      })
    ).toMatchSnapshot();

    expect(find.mock.calls).toMatchSnapshot();
  });

  test("saveSearchResult should make the right query", async () => {
    const save = jest.fn(() => [{ data: "yay" }]);
    const repository = {
      save
    };

    expect(
      await saveSearchResult(
        (repository as unknown) as Repository<EntitySearchResult>,
        // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
        // @ts-ignore
        [{ id: 123 }, { id: 456 }],
        { id: 890 }
      )
    ).toMatchSnapshot();

    expect(save.mock.calls).toMatchSnapshot();
  });

  test("fetchSearchResult should make the right query", async () => {
    const find = jest.fn(() => [{ data: "yay" }]);
    const repository = {
      find
    };

    expect(
      await fetchSearchResult(
        (repository as unknown) as Repository<EntitySearchResult>,
        {
          id: 890,
          album: "album",
          source: "source",
          orderBy: "orderBy",
          searchResults: [],
          filterBy: ""
        },
        0,
        null
      )
    ).toMatchSnapshot();

    expect(find.mock.calls).toMatchSnapshot();
  });
});
