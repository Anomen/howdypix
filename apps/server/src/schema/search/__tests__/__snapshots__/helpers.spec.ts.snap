// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchHelpers doSearch should make the right query 1`] = `
Array [
  Object {
    "data": "yay",
  },
]
`;

exports[`searchHelpers doSearch should make the right query 2`] = `
Array [
  Array [
    Object {
      "order": Object {
        "birthtime": "ASC",
      },
      "where": Object {
        "dir": "album",
        "source": "source",
      },
    },
  ],
]
`;

exports[`searchHelpers fetchSearchResult should make the right query 1`] = `
Array [
  Object {
    "data": "yay",
  },
]
`;

exports[`searchHelpers fetchSearchResult should make the right query 2`] = `
Array [
  Array [
    Object {
      "order": Object {
        "order": "ASC",
      },
      "relations": Array [
        "photo",
      ],
      "where": Object {
        "search": Object {
          "album": "album",
          "filterBy": "",
          "id": 890,
          "orderBy": "orderBy",
          "searchResults": Array [],
          "source": "source",
        },
      },
    },
  ],
]
`;

exports[`searchHelpers findSavedSearch should make the right query 1`] = `
Object {
  "data": "yay",
}
`;

exports[`searchHelpers findSavedSearch should make the right query 2`] = `
Array [
  Array [
    Object {
      "where": Object {
        "album": "album",
        "filterBy": null,
        "orderBy": "DATE_ASC",
        "source": "source",
      },
    },
  ],
]
`;

exports[`searchHelpers saveNewSearch should make the right query 1`] = `
Object {
  "data": "yay",
}
`;

exports[`searchHelpers saveNewSearch should make the right query 2`] = `
Array [
  Array [
    Search {
      "album": "album",
      "filterBy": undefined,
      "id": undefined,
      "orderBy": "DATE_ASC",
      "searchResults": undefined,
      "source": "source",
    },
  ],
]
`;

exports[`searchHelpers saveSearchResult should make the right query 1`] = `
Array [
  Array [
    Object {
      "data": "yay",
    },
  ],
  Array [
    Object {
      "data": "yay",
    },
  ],
]
`;

exports[`searchHelpers saveSearchResult should make the right query 2`] = `
Array [
  Array [
    SearchResult {
      "id": undefined,
      "order": 0,
      "photo": Object {
        "id": 123,
      },
      "photoId": undefined,
      "search": Object {
        "id": 890,
      },
      "searchId": undefined,
    },
  ],
  Array [
    SearchResult {
      "id": undefined,
      "order": 1,
      "photo": Object {
        "id": 456,
      },
      "photoId": undefined,
      "search": Object {
        "id": 890,
      },
      "searchId": undefined,
    },
  ],
]
`;
